immer/algorithm.hpp

```diff
@@ -221,6 +240,22 @@ bool all_of(Iter first, Iter last, Pred p)
     });
 }
 
+/*!
+ * Equivalent of the `std::uninitialized_copy` applied to the
+ * sentinel-delimited forward range @f$ [first, last) @f$
+ */
+template <typename SourceIter, typename Sent, typename SinkIter,
+          std::enable_if_t
+          <detail::compatible_sentinel<SourceIter,Sent>
+           and detail::is_forward_iterator<SinkIter>, bool> = true>
+auto uninitialized_copy(SourceIter first, Sent last, SinkIter d_first){
+    while(first != last){
+        *d_first++ = *first;
+        ++first;
+    }
+    return d_first;
+}
+
```

Comment:
Same as above: this could be done more efficient in the context of `immer` data-structures.

Reply:
Also, this implementation of `uninitialized_copy` is not correct, because it does not take into account exceptions: if an exception is thrown you have to roll-back, by destructing the objects that have already been copied. This might be the reason why your branch is introducing memory leaks. Still, I would suggest to avoid adding these algorithms here.